### PR TITLE
enables retry on connection drop

### LINKED ISSUE #313: ClientException thrown when tcp connection is closed
Expected behavior
When a underlying connection okhttp is using is closed okhttp should reopen a new connection without throwing an exception.
Actual behavior
ClientException is thrown by the graph sdk.
Underlying connection okhttp is handling is closed causing a EOFException to be thrown. This eventually bubbles up to the sdk library and is caught and rethrown as a ClientException.
Steps to reproduce the behavior
Make a number of calls to the graph sdk consistantly over a period of 15-30 minutes and the error should be seen.
This looks to be caused by setting okBuilder.retryOnConnectionFailure(false); in CoreHttpProvider. I can't see why this was done but it happened when RetryHandler functionality was added from msgraph-sdk-java-core. Since RetryHandler only retries status code responses of 429 503 504 and doesn't actually do anything about exeptions this should be set to true.
There's no way to override this setting the only workaround I've found is to add retry logic into any code that makes a call with the sdk or inject a 'Connection: close' header into every request made making okhttp open a new connection everytime.
relevant info: square/okhttp/issues/1873
com.microsoft.graph.core.ClientException: Error during http request
com.microsoft.graph.http.CoreHttpProvider.sendRequestInternal(CoreHttpProvider.java:422) ~[microsoft-graph-1.6.0.jar!/:na]\tat com.microsoft.graph.http.CoreHttpProvider.send(CoreHttpProvider.java:204) ~[microsoft-graph-1.6.0.jar!/:na]
 com.microsoft.graph.http.CoreHttpProvider.send(CoreHttpProvider.java:184) ~[microsoft-graph-1.6.0.jar!/:na]
 com.microsoft.graph.http.BaseCollectionRequest.send(BaseCollectionRequest.java:89) ~[microsoft-graph-1.6.0.jar!/:na]
 com.microsoft.graph.requests.extensions.UserCollectionRequest.get(UserCollectionRequest.java:52) ~[microsoft-graph-1.6.0.jar!/:na]
.......
 org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61) [tomcat-embed-core-9.0.24.jar!/:9.0.24]
 java.lang.Thread.run(Thread.java:748) [na:1.8.0_222]
Caused by: java.io.IOException: unexpected end of stream on https://graph.microsoft.com/...
 okhttp3.internal.http1.Http1ExchangeCodec.readResponseHeaders(Http1ExchangeCodec.java:236) ~[okhttp-3.14.4.jar!/:na]
 okhttp3.internal.connection.Exchange.readResponseHeaders(Exchange.java:115) ~[okhttp-3.14.4.jar!/:na]
 okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.java:94) ~[okhttp-3.14.4.jar!/:na]
 okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142) ~[okhttp-3.14.4.jar!/:na]
 okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:43) ~[okhttp-3.14.4.jar!/:na]
 okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142) ~[okhttp-3.14.4.jar!/:na]
 okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117) ~[okhttp-3.14.4.jar!/:na]
 okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:94) ~[okhttp-3.14.4.jar!/:na]
 okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142) ~[okhttp-3.14.4.jar!/:na]
 okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117) ~[okhttp-3.14.4.jar!/:na]
 okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93) ~[okhttp-3.14.4.jar!/:na]
 okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142) ~[okhttp-3.14.4.jar!/:na]
 okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:88) ~[okhttp-3.14.4.jar!/:na]
 okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142) ~[okhttp-3.14.4.jar!/:na]
 okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117) ~[okhttp-3.14.4.jar!/:na]
 com.microsoft.graph.httpcore.TelemetryHandler.intercept(TelemetryHandler.java:35) ~[microsoft-graph-core-1.0.0.jar!/:na]
 okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142) ~[okhttp-3.14.4.jar!/:na]
 okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117) ~[okhttp-3.14.4.jar!/:na]
 com.microsoft.graph.httpcore.RedirectHandler.intercept(RedirectHandler.java:123) ~[microsoft-graph-core-1.0.0.jar!/:na]
 okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142) ~[okhttp-3.14.4.jar!/:na]
 okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117) ~[okhttp-3.14.4.jar!/:na]
 com.microsoft.graph.httpcore.RetryHandler.intercept(RetryHandler.java:140) ~[microsoft-graph-core-1.0.0.jar!/:na]
 okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142) ~[okhttp-3.14.4.jar!/:na]
 okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117) ~[okhttp-3.14.4.jar!/:na]
 com.microsoft.graph.httpcore.AuthenticationHandler.intercept(AuthenticationHandler.java:31) ~[microsoft-graph-core-1.0.0.jar!/:na]
 okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142) ~[okhttp-3.14.4.jar!/:na]
 okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117) ~[okhttp-3.14.4.jar!/:na]
 okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:229) ~[okhttp-3.14.4.jar!/:na]
 okhttp3.RealCall.execute(RealCall.java:81) ~[okhttp-3.14.4.jar!/:na]
 com.microsoft.graph.http.CoreHttpProvider.sendRequestInternal(CoreHttpProvider.java:356) ~[microsoft-graph-1.6.0.jar!/:na]
... 105 common frames omitted
Caused by: java.io.EOFException: 
 not found: limit=0 content=…
 okio.RealBufferedSource.readUtf8LineStrict(RealBufferedSource.java:240) ~[okio-1.17.2.jar!/:na]
 okhttp3.internal.http1.Http1ExchangeCodec.readHeaderLine(Http1ExchangeCodec.java:242) ~[okhttp-3.14.4.jar!/:na]\tat okhttp3.internal.http1.Http1ExchangeCodec.readResponseHeaders(Http1ExchangeCodec.java:213) ~[okhttp-3.14.4.jar!/:na]
\t... 134 common frames omitted
[AB#6038](https://microsoftgraph.visualstudio.com/0985d294-5762-4bc2-a565-161ef349ca3e/_workitems/edit/6038)

### PR DESCRIPTION
**Pull request by baywet**
Fixes #313

### Changes proposed in this pull request
-

### Other links

-

### CODE REVIEW

File: src/main/java/com/microsoft/graph/http/CoreHttpProvider.java [hunk 1]
```diff
@@ -356,20 +356,22 @@ public MediaType contentType() {
...
+															@Override
+															public Request authenticateRequest(Request request) {
+																return request;
+															}
+														}).newBuilder()
+										.connectTimeout(connectionConfig.getConnectTimeout(), TimeUnit.MILLISECONDS)
+										.readTimeout(connectionConfig.getReadTimeout(), TimeUnit.MILLISECONDS)
+										.followRedirects(false)
+										.protocols(Arrays.asList(Protocol.HTTP_1_1)) //https://stackoverflow.com/questions/62031298/sockettimeout-on-java-11-but-not-on-java-8
+										.build();
```
Comment (rocketraman) [inline src/main/java/com/microsoft/graph/http/CoreHttpProvider.java:376]: The indentation is messed up.

**Review by MIchaelMainer — DISMISSED** [review]
Do you always see a dropped connection + retry in the test? 

https://github.com/microsoftgraph/msgraph-sdk-java/pull/473/files#diff-6709b7f7a4c3667aae557143946c5ab2R180

**Comment by baywet** [comment]
@MIchaelMainer this actually fixes 3 problems:
- Okhttp on jre11+ initiates the connection using HTTP/2 by default instead of http/1.1 for jre8. Our services for some reason send a connection reset when they see HTTP/2 coming. -> locks to http/1.1
- Our retry handler only retries on receiving responses  (429, 504,...) but it doesn't retry on connection drop/reset etc (for some reason AGS sends a connection reset when it gets hot). -> enables retry connection from the http lib.
- Our retry handler was never integration tested -> adds a integration test that tests heavy load working. (which also allowed me to debug the other two cases)

File: src/main/java/com/microsoft/graph/http/CoreHttpProvider.java (hunk 1 above)
Comment (baywet) [reply src/main/java/com/microsoft/graph/http/CoreHttpProvider.java]: fixed

**Review by MIchaelMainer — APPROVED** [review]

**Comment by yschimke** [comment]
It's worth testing this with OkHttp 4.10.0-RC1

https://square.github.io/okhttp/changelog/#version-4100-rc1

> Fix: Attempt to read the response body even if the server canceled the request. This will cause some calls to return nice error codes like HTTP/1.1 429 Too Many Requests instead of transport errors like SocketException: Connection reset and StreamResetException: stream was reset: CANCEL.

**Comment by baywet** [comment]
Hey @yschimke 
Thanks for letting us know! I logged #554 as we plan to upgrade ok http with that version

**Pull request merged**